### PR TITLE
feat: create page that shows previously set sales offer packages

### PR DESCRIPTION
**Pull request by gilesdrinkall**
# Description

To add a page that allows users to select previously set sales offer packages.

# Testing instructions

Can be tested locally using the fdbt-dev repo.

# Type of change

-   [x] feat - A new feature
-   [ ] fix - A bug fix
-   [ ] docs - Documentation only changes
-   [ ] codestyle - Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
-   [ ] refactor - A code change that neither fixes a bug nor adds a feature
-   [ ] perf - A code change that improves performance
-   [ ] test - Adding missing tests or correcting existing tests
-   [ ] build - Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
-   [ ] ci - Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
-   [ ] chore - Other changes that don't modify src or test files
-   [ ] revert - Reverts a previous commit
-   [ ] content - Changes to content or copy

# Checklist:

-   [x] Able to run pr locally
-   [ ] Lighthouse performed (Accessibility 90+ minimum)
-   [x] Followed acceptance criteria
-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [ ] I have made corresponding changes to the documentation if applicable
-   [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/interfaces/index.ts [hunk 1]
```diff
@@ -86,3 +90,136 @@ export type DocumentContextWithSession = DocumentContext & {
 };
 
 export type IncomingMessageWithSession = IncomingMessage & Session;
+
+export interface SessionContents {
+    body: any;
+}
```
Comment (dannydaviesIW) [inline src/interfaces/index.ts:96]: We should remove this `SessionContents` as its no longer used

File: src/interfaces/index.ts [hunk 2]
```diff
@@ -32,6 +32,10 @@ export interface PassengerDetails {
     proof?: string[];
 }
 
+export interface SelectSalesOfferPackageWithError {
+    errorMessage: string;
+}
```
Comment (dannydaviesIW) [inline src/interfaces/index.ts:37]: Should this `SelectSalesOfferPackageWithError` interface live on the page/api where it is used? I think most of these kinds of interfaces (which are only used for one page/api) are stored this way

File: src/interfaces/index.ts [hunk 3]
```diff
@@ -86,3 +90,136 @@ export type DocumentContextWithSession = DocumentContext & {
...
+    productValidity?: string;
+}
+
+export interface PeriodProductData {
+    products: ProductDetails[];
+}
+
+export interface FlatFareProductData {
+    products: FlatFareProductDetails[];
+}
```
Comment (dannydaviesIW) [inline src/interfaces/index.ts:119]: We can remove `PeriodProductData` and `FlatFareProductData` as we didnt end up using these

File: src/interfaces/index.ts [hunk 4]
```diff
@@ -86,3 +90,136 @@ export type DocumentContextWithSession = DocumentContext & {
...
+
+export interface ProductData {
+    products: Product[];
+}
+
+export interface PeriodExpiryWithErrors {
+    errorMessage: string;
+}
```
Comment (dannydaviesIW) [inline src/interfaces/index.ts:127]: Same here as with `SelectSalesOfferPackageWithError`

File: src/interfaces/index.ts (hunk 2 above)
Comment (dannydaviesIW) [reply src/interfaces/index.ts:37]: Also, should this not be 
```
export interface SelectSalesOfferPackageWithError {
    errors: ErrorInfo[];
}
```

File: src/pages/api/inboundMatching.ts [hunk 5]
```diff
@@ -89,55 +22,24 @@ export default async (req: NextApiRequest, res: NextApiResponse): Promise<void>
             throw new Error('No service or userfarestages info found');
         }
```
Comment (dannydaviesIW) [inline src/pages/api/inboundMatching.ts:23]: We should get rid of the `isCookiesUuidMatch()` check as it has caused issues already on the matching API

File: src/pages/api/multipleProductValidity.ts [hunk 6]
```diff
@@ -125,7 +124,7 @@ export default async (req: NextApiRequest, res: NextApiResponse): Promise<void>
             throw new Error('Could not extract the user email address from their ID token');
         }
 
-        const multipleProductPeriod: DecisionData = {
+        const multipleProductPeriod = {
```
Comment (dannydaviesIW) [inline src/pages/api/multipleProductValidity.ts:127]: Should we be replacing this type we've removed?

File: src/pages/api/outboundMatching.ts [hunk 7]
```diff
@@ -38,37 +22,31 @@ export default async (req: NextApiRequest, res: NextApiResponse): Promise<void>
...
-
-        await putOutboundMatchingDataInS3(matchedFareZones, uuid);
-
-        setCookieOnResponseObject(MATCHING_COOKIE, JSON.stringify({ outbound: { error: false } }), req, res);
-
+        updateSessionAttribute(req, MATCHING_ATTRIBUTE, {
+            service,
+            userFareStages,
+            matchingFareZones: matchedFareZones,
+        });
```
Comment (dannydaviesIW) [inline src/pages/api/outboundMatching.ts:49]: We should create a variable with this information prior to this function call. That way we can assert a type and check that we are formatting the data correctly when we call `updateSessionAttribute`

File: src/pages/api/selectSalesOfferPackage.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                redirectTo(res, '/thankyou');
+                return;
+            }
+        }
+        if (fareType === 'flatFare') {
+            userDataJson = getFlatFareTicketJson(req, res);
+            await putUserDataInS3(userDataJson, uuid);
+            redirectTo(res, '/thankyou');
+            return;
+        }
```
Comment (dannydaviesIW) [inline src/pages/api/selectSalesOfferPackage.ts:79]: We could make this block of if statements a bit shorter and more readable by only setting `userDataJson` inside - by removing the below from each if block and have it just once at the end
```
await putUserDataInS3(userDataJson, uuid);
            redirectTo(res, '/thankyou');
            return;
```
That way we would replace 15 lines of code with 3

File: src/pages/selectSalesOfferPackage.tsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+import { getSalesOfferPackagesByNocCode } from '../data/auroradb';
+import { SalesOfferPackage, CustomAppProps, ErrorInfo, NextPageContextWithSession } from '../interfaces';
+import { getNocFromIdToken } from '../utils';
+import CsrfForm from '../components/CsrfForm';
+import { redirectTo } from './api/apiUtils';
+import { getSessionAttribute } from '../utils/sessions';
+
+const pageTitle = 'Select Sales Offer Package - Fares Data Build Tool';
+const pageDescription = 'Sales Offer Package selection page of the Fares Data Build Tool';
+const errorId = 'sales-offer-package-error';
```
Comment (dannydaviesIW) [inline src/pages/selectSalesOfferPackage.tsx:15]: Does this errorId actually link to any ID on the page?

File: src/interfaces/index.ts (hunk 1 above)
Comment (gilesdrinkall) [reply src/interfaces/index.ts:96]: Now removed.

File: src/interfaces/index.ts (hunk 2 above)
Comment (gilesdrinkall) [reply src/interfaces/index.ts:37]: I have moved the interface to SelectSalesOfferPackage.ts. I haven't change it to comprise a ErrorInfo[] under a key of errors as we only have a single error message on SelectSalesOfferPackage.ts.

File: src/interfaces/index.ts (hunk 4 above)
Comment (gilesdrinkall) [reply src/interfaces/index.ts:127]: I have moved the interface to periodValidity.ts. I haven't change it to comprise a ErrorInfo[] under a key of errors as we only have a single error message on periodValidity.ts.

File: src/interfaces/index.ts (hunk 3 above)
Comment (gilesdrinkall) [reply src/interfaces/index.ts:119]: Now removed.

File: src/pages/api/inboundMatching.ts (hunk 5 above)
Comment (gilesdrinkall) [reply src/pages/api/inboundMatching.ts:23]: Removed.

File: src/pages/api/multipleProductValidity.ts (hunk 6 above)
Comment (gilesdrinkall) [reply src/pages/api/multipleProductValidity.ts:127]: This code will be removed entirely by the work being done on 535, so no need to annotate with a type at this stage.

File: src/pages/api/outboundMatching.ts (hunk 7 above)
Comment (gilesdrinkall) [reply src/pages/api/outboundMatching.ts:49]: Created a variable and a type (MatchingValues).

File: src/pages/selectSalesOfferPackage.tsx (hunk 9 above)
Comment (gilesdrinkall) [reply src/pages/selectSalesOfferPackage.tsx:15]: No

File: src/data/auroradb.ts [hunk 10]
```diff
@@ -165,6 +165,33 @@ export const getServicesByNocCode = async (nocCode: string): Promise<ServiceType
     }
 };
 
+export const getSalesOfferPackagesByNocCode = async (nocCode: string): Promise<SalesOfferPackage[]> => {
+    const nocCodeParameter = replaceIWBusCoNocCode(nocCode);
```
[the remark this answers is not in the document]
Comment (gilesdrinkall) [reply src/data/auroradb.ts:169]: Have now removed function call that replaces NocCode.

File: src/pages/api/apiUtils/index.ts [hunk 11]
```diff
@@ -143,3 +180,318 @@ export const validateNewPassword = (
...
+): Promise<void> => {
+    await putStringInS3(
+        MATCHING_DATA_BUCKET_NAME,
+        `${data.nocCode}/${data.type}/${uuid}_${Date.now()}.json`,
+        JSON.stringify(data),
+        'application/json; charset=utf-8',
+    );
+};
+
+export const getSingleTicketJson = (req: NextApiRequestWithSession, res: NextApiResponse): SingleTicket => {
```
[the remark this answers is not in the document]
Comment (gilesdrinkall) [reply src/pages/api/apiUtils/index.ts:215]: Now moved to a new file called 'userData.ts' in the 'apiUtils' folder.

**Pull request merged**